crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -185,7 +179,45 @@ impl Solution {
     /// Check that the sum of tokens entering the settlement is not less than
     /// the sum of tokens exiting the settlement.
     fn verify_asset_flow(&self) -> Result<(), VerificationError> {
-        Ok(())
+        let mut flow: HashMap<eth::TokenAddress, num::BigInt> = Default::default();
+
+        // Interaction inputs represent flow into the contract, i.e. positive flow.
+        for input in self
+            .interactions
+            .iter()
+            .flat_map(|interaction| interaction.inputs())
+        {
+            *flow.entry(input.token).or_default() += util::conv::u256::to_big_int(input.amount);
+        }
+
+        // Interaction outputs represent flow out of the contract, i.e. negative flow.
+        for output in self
+            .interactions
+            .iter()
+            .flat_map(|interaction| interaction.outputs())
+        {
+            *flow.entry(output.token).or_default() -= util::conv::u256::to_big_int(output.amount);
+        }
+
+        // For trades, the input and output amounts are determined by the type of trade.
+        for trade in self.trades.iter().filter_map(|trade| match trade {
+            Trade::Fulfillment(trade) => Some(trade),
+            // JIT trades are "fake" trades inserted by the solver to improve the solution score,
+            // and hence should not be considered for asset flow verification.
+            Trade::Jit(..) => None,
```

Comment:
I'm actually no longer sure `JIT` orders need a special treatment with this kind of asset flow test. cc @nlordell?

Reply:
I think they definitely need to be considered. They are an important part of token conservation.

Take, for example, an order that gets settled directly with a JIT order. It would have a +ve flux of sell token going into the settlement contract, and a -ve flux of buy token out of the settlement contract. This would lead to breaking token conservation (which shouldn't be the case, and wouldn't be if you also consider the JIT order's amounts).

Reply:
Yes, I know. That's why I wrote "I'm actually no longer sure...".